client/lib/route/path.ts

```diff
@@ -17,46 +17,47 @@ export function getSiteFragment( path: URLString ): SiteSlug | SiteId | false {
 	const basePath = path.split( '?' )[ 0 ];
 	const pieces = basePath.split( '/' );
 
-	// There are 2 URL positions where we should look for the site fragment:
-	// last (most sections) and second-to-last (post ID is last in editor)
-
-	// Avoid confusing the receipt ID for the site ID in domain-only checkouts.
-	if ( 0 === basePath.indexOf( '/checkout/thank-you/no-site/' ) ) {
+	// Some paths include a receipt or subscription ID that can be mistaken for a site ID.
+	// Always return false for these specific paths.
+	if (
+		// Avoid confusing the receipt ID for the site ID in domain-only checkouts.
+		0 === basePath.indexOf( '/checkout/thank-you/no-site/' ) ||
+		// Avoid confusing the subscription ID for the site ID in gifting checkouts.
+		( basePath.includes( '/gift/' ) && basePath.startsWith( '/checkout/' ) )
+	) {
 		return false;
 	}
 
-	// Avoid confusing the subscription ID for the site ID in gifting checkouts.
-	if ( basePath.includes( '/gift/' ) && basePath.startsWith( '/checkout/' ) ) {
-		return false;
-	}
+	// By default there are two URL positions where we should look for the site fragment:
+	// last (most sections) and second-to-last (post ID is last in editor).
+	// searchPositions defines the second-to-last and last index positions to search. Order matters.
+	let searchPositions = [ pieces.length - 2, pieces.length - 1 ];
 
-	// In some paths, the site fragment could also be in third position.
+	// There are exceptions. In some paths the site fragment is in the third position.
 	// e.g. /me/purchases/example.wordpress.com/foo/bar
 	if (
 		0 === basePath.indexOf( '/me/purchases/' ) ||
 		0 === basePath.indexOf( '/checkout/thank-you/' )
 	) {
-		const piece = pieces[ 3 ]; // 0 is the empty string before the first `/`
-		if ( piece && -1 !== piece.indexOf( '.' ) ) {
-			return piece;
-		}
-		const numericPiece = parseInt( piece, 10 );
-		if ( Number.isSafeInteger( numericPiece ) ) {
-			return numericPiece;
-		}
```

Comment:
The code here is a duplication of the code in the `for` loop below, but missing some recent updates. To prevent duplicating this code again for our "upsell path" exception but looking in the 2nd URL position, I've refactored the `getSiteFragment` function to define the `searchPositions` we'd like to look for the slug or ID in. Then at the bottom of the function we loop through `searchPositions` looking for the slug or ID.

Reply:
I'm leaning pretty heavy on the unit tests for regression testing. Eyeballing those for any missing permutations would be good. 🤔 I suppose we could also cruise WordPress.com watching `calypso_page_view` in the console for the correct `path`.

Reply:
> Eyeballing those for any missing permutations would be good.

I'm not seeing any issues but more 👀 seem like a good idea. @tyxla @Automattic/kitkat Any chance you can take a look at the change here and whether we have sufficient test coverage?

Reply:
Thank you! 🙌